src/realm/set.hpp

```diff
@@ -0,0 +1,455 @@
+
+    /// Find the index of a value in the set, or `size_t(-1)` if it is not in the set.
+    size_t find(T value) const;
+
+    /// Erase an element from the set, returning the index at which it was removed, or `size_t(-1)` if it did
+    /// not exist.
+    size_t erase(T value);
+
+    // Overriding members of CollectionBase:
+    Mixed min(size_t* return_ndx = nullptr) const final;
+    Mixed max(size_t* return_ndx = nullptr) const final;
+    Mixed sum(size_t* return_cnt = nullptr) const final;
+    Mixed avg(size_t* return_cnt = nullptr) const final;
+    void sort(std::vector<size_t>& indices, bool ascending = true) const final;
+    void distinct(std::vector<size_t>& indices, util::Optional<bool> sort_order = util::none) const final;
+
+    // Overriding members of SetBase:
+    void insert_null() override;
+    void erase_null() override;
+    void insert_any(Mixed value) override;
+    void erase_any(Mixed value) override;
+    void clear() override;
+
+private:
+    using Collection<T, SetBase>::m_valid;
+    using Collection<T, SetBase>::m_obj;
+
+    void create()
+    {
+        m_tree->create();
+        m_valid = true;
+    }
+
+    bool update_if_needed()
+    {
+        if (m_obj.update_if_needed()) {
+            return this->init_from_parent();
+        }
+        return false;
+    }
+    void ensure_created()
+    {
+        if (!m_valid && m_obj.is_valid()) {
+            create();
+        }
+    }
+};
+
+/// Compare set elements.
+///
+/// We cannot use `std::less<>` directly, because the ordering of set elements
+/// impacts the file format. For primitive types this is trivial (and can indeed
+/// be just `std::less<>`), but for example `Mixed` has specialized comparison
+/// that defines equality of numeric types.
+template <class T>
+struct SetElementLessThan {
+    bool operator()(const T& a, const T& b) const noexcept
+    {
+        // CAUTION: This routine is technically part of the file format, because
+        // it determines the storage order of Set elements.
+        return a < b;
+    }
+};
+
+template <class T>
+struct SetElementEquals {
+    bool operator()(const T& a, const T& b) const noexcept
+    {
+        // CAUTION: This routine is technically part of the file format, because
+        // it determines the storage order of Set elements.
+        return a == b;
+    }
+};
+
+template <>
+struct SetElementLessThan<Mixed> {
+    bool operator()(const Mixed& a, const Mixed& b) const noexcept
+    {
+        // CAUTION: This routine is technically part of the file format, because
+        // it determines the storage order of Set elements.
+
+        if (a.is_null() != b.is_null()) {
+            // If a is NULL but not b, a < b.
+            return a.is_null();
+        }
+        else if (a.is_null()) {
+            // NULLs are equal.
+            return false;
+        }
+
+        if (a.get_type() != b.get_type()) {
+            return a.get_type() < b.get_type();
+        }
+
```

Comment:
Because I wanted to avoid relying on `Mixed::compare()` for the implementation of Set, but we may have to do that anyway because we want to match MongoDB comparisons for `$addToSet`.

I'll have to double-check that the comparisons match. I've create this issue to do that in a follow-up PR: https://jira.mongodb.org/projects/RCORE/issues/RCORE-359